packages/react-components/react-combobox/src/components/Combobox/useCombobox.tsx

```diff
@@ -18,22 +20,126 @@ import type { ComboboxProps, ComboboxState } from './Combobox.types';
+    setHasFocus(true);
+
+    triggerNativeProps.onFocus?.(ev);
+  };
+
+  const getSearchString = (inputValue: string): string => {
+    // if there are commas in the value string, take the text after the last comma
+    const searchString = inputValue.split(',').pop();
+
+    return searchString?.trim().toLowerCase() || '';
+  };
+
+  // set active option and selection based on typing
+  const getOptionFromInput = (inputValue: string): OptionValue | undefined => {
+    const searchString = getSearchString(inputValue);
+
+    if (searchString.length === 0) {
+      return;
+    }
+
+    const matcher = (optionValue: string) => optionValue.toLowerCase().indexOf(searchString) === 0;
+    const matches = getOptionsMatchingValue(matcher);
+
+    // return first matching option after the current active option, looping back to the top
+    if (matches.length > 1 && activeOption) {
+      const startIndex = getIndexOfId(activeOption.id);
+      const nextMatch = matches.find(option => getIndexOfId(option.id) >= startIndex);
+      return nextMatch ?? matches[0];
+    }
+
+    return matches[0] ?? undefined;
+  };
+
+  /* Handle typed input */
+
+  // reset any typed value when an option is selected
+  baseState.selectOption = (ev: SelectionEvents, option: OptionValue) => {
+    setValue(undefined);
+    selectOption(ev, option);
+  };
+
+  const onBlur = (ev: React.FocusEvent<HTMLInputElement>) => {
+    // handle selection and updating value if allowFreeform is false
+    if (!baseState.open && !allowFreeform) {
+      // select matching option, if the value fully matches
+      if (value && activeOption && getSearchString(value) === activeOption?.value.toLowerCase()) {
+        baseState.selectOption(ev, activeOption);
+      }
+
+      // reset typed value when the input loses focus while collapsed, unless allowFreeform is true
+      setValue(undefined);
+    }
+
+    setHasFocus(false);
+    triggerNativeProps.onBlur?.(ev);
+  };
+
+  baseState.setOpen = (ev, newState: boolean) => {
+    if (!newState && !allowFreeform) {
+      setValue(undefined);
+    }
+
+    setOpen(ev, newState);
+  };
+
+  // update value and active option based on input
+  const onChange = (ev: React.ChangeEvent<HTMLInputElement>) => {
+    const inputValue = ev.target.value;
+    // update uncontrolled value
+    baseState.setValue(inputValue);
+
+    // handle updating active option based on input
+    const matchingOption = getOptionFromInput(inputValue);
+    setActiveOption(matchingOption);
+
+    // clear selection for single-select if the input value no longer matches the selection
+    if (
+      !multiselect &&
+      selectedOptions.length === 1 &&
+      (inputValue.length < 1 || selectedOptions[0].indexOf(inputValue) !== 0)
+    ) {
+      clearSelection(ev);
+    }
+
+    triggerNativeProps.onChange?.(ev);
+  };
+
+  // resolve input and listbox slot props
   const triggerShorthand = resolveShorthand(props.input, {
```

Comment:
I honestly spent like 10 minutes trying to figure out the names for these 😂. I know the return from `resolveShorthand` isn't shorthand, but the problem is that a few lines down, I already use `triggerSlot` for the return of `useTriggerListboxSlots` :P.

I'll think about it a bit more.

Reply:
You could use `let` instead of `const` so you can reassign the same variable when you call useTriggerListboxSlots.
```ts
let trigger = resolveShorthand(/*...*/);

// ...

[trigger, listbox] = useComboboxPopup(/*...*/);
[trigger, listbox] = useTriggerListboxSlots(/*...*/);
```